BEAMS3D/Sources/beams3d_beam_density.f90

```diff
@@ -0,0 +1,121 @@
+      LOGICAL, DIMENSION(:,:,:,:), ALLOCATABLE :: lgrid
+      DOUBLE PRECISION, DIMENSION(:), ALLOCATABLE :: xl,yl,zl,rl,sl,pl
+#if defined(MPI_OPT)
+      INTEGER :: MPI_COMM_LOCAL
+      INTEGER :: mystart, mypace
+#endif
+!-----------------------------------------------------------------------
+!     Begin Subroutine
+!-----------------------------------------------------------------------
+
+      ! Allocate the Grid
+      CALL mpialloc(BEAM_DENSITY, nbeams, nr, nphi, nz, myid_sharmem, 0, MPI_COMM_SHARMEM, win_BEAM_DENSITY)
+      IF (myid_sharmem == 0) BEAM_DENSITY = 0
+      ALLOCATE(lgrid(nbeams,nr,nphi,nz))
+      nl = nr + nr/2 + 1
+      ALLOCATE(xl(nl),yl(nl),zl(nl),rl(nl),pl(nl),sl(nl))
+      FORALL(s=1:nl) sl(s) = DBLE(s-1)/DBLE(nl-1)
+
+
+      IF (lverb) WRITE(6,'(A)') '----- BEAM Density Calc. -----'
+      hr2 = 0.5*hr(1)
+      hz2 = 0.5*hz(1)
+      hp2 = 0.5*hp(1)
+      ! Loop over particles
+      DO l = mystart_save, myend_save
+         x0 = R_lines(0,l)*cos(PHI_lines(0,l))
+         y0 = R_lines(0,l)*sin(PHI_lines(0,l))
+         z0 = Z_lines(0,l)
+         x1 = R_lines(1,l)*cos(PHI_lines(1,l))
+         y1 = R_lines(1,l)*sin(PHI_lines(1,l))
+         z1 = Z_lines(1,l)
+         m  = Beam(l)
+         ! Because we're in cylindrical space we just need to brute force this
+         ! We have a chord with particles/s (weight) going along it.
+         ! Since the velocity is constant then the TOF is just L/v = t
+         ! The the total number of particles is just n=W*t = W*L/v
+         ! then we jsut need to calc the volume of the voxel to get particles/m^3
+         d  = SQRT((x1-x0)**2 + (y1-y0)**2 + (z1-z0)**2)
+         denbeam = weight(l)*d/vll_lines(0,l) ! This is the total number of particles
+         xl = x0 + sl*(x1-x0)
+         yl = y0 + sl*(y1-y0)
+         zl = z0 + sl*(z1-z0)
+         rl = sqrt(xl*xl+yl*yl)
+         pl = atan2(yl,xl)
+         pl = MODULO(pl,phimax)
+         LGRID = .FALSE.
+         DO s = 1, nl
+            ! Find gridpoint but use half grid
+            i = MIN(MAX(COUNT(raxis-hr2 < rl(s)),1),nr)
+            j = MIN(MAX(COUNT(phiaxis-hp2 < pl(s)),1),nphi)
+            k = MIN(MAX(COUNT(zaxis-hz2 < zl(s)),1),nz)
+            LGRID(m,i,j,k) = .true.
+         END DO
+         WHERE(lgrid) BEAM_DENSITY = BEAM_DENSITY + denbeam
+      END DO
+      DEALLOCATE(lgrid)
+      DEALLOCATE(xl,yl,zl,rl,pl,sl)
+
+      ! Fix edges which are double counts
+      IF (myid_sharmem == 0) THEN
+         BEAM_DENSITY(:,1,:,:) = 0
+         BEAM_DENSITY(:,nr,:,:) = 0
+         BEAM_DENSITY(:,:,:,1) = 0
+         BEAM_DENSITY(:,1,:,nz) = 0
+      ENDIF
+
+
+      ! Barrier so  calculation is done
+#if defined(MPI_OPT)
+      CALL MPI_BARRIER(MPI_COMM_SHARMEM,ierr_mpi)
+#endif
+      ! Now Divide by the grid volumes
+      CALL MPI_CALC_MYRANGE(MPI_COMM_SHARMEM, 1, (nr-1)*(nphi-1)*(nz-1), mystart, myend)
+      DO s = mystart,myend
+         i = MOD(s-1,nr-1)+1
+         j = MOD(s-1,(nr-1)*(nphi-1))
```

Comment:
We should ensure that the phi grid isnt too coarse to give an artificially low density, i.e. when dphi*raxis(nr) is larger than the beam diameter, the volume average is lowered.

Reply:
Added an warning message if this is an issue.